openrtb_ext/request.go

```diff
@@ -71,14 +71,186 @@ func (ert *ExtRequestTargeting) UnmarshalJSON(b []byte) error {
 }
 
 // PriceGranularity defines the allowed values for bidrequest.ext.prebid.targeting.pricegranularity
-type PriceGranularity string
-
-const (
-	PriceGranularityLow    PriceGranularity = "low"
-	PriceGranularityMedium PriceGranularity = "medium"
-	// Seems that PBS was written with medium = "med", so hacking that in
-	PriceGranularityMedPBS PriceGranularity = "med"
-	PriceGranularityHigh   PriceGranularity = "high"
-	PriceGranularityAuto   PriceGranularity = "auto"
-	PriceGranularityDense  PriceGranularity = "dense"
-)
+type PriceGranularity []GranularityRange
+
+type granularityRangeRaw struct {
+	Precision *int    `json:"precision,omitempty"`
+	Min       float64 `json:"min,omitempty"`
+	Max       float64 `json:"max"`
+	Increment float64 `json:"increment"`
+}
+
+// GranularityRange struct defines a range of prices used by PriceGranularity
+type GranularityRange struct {
+	Precision int
+	Min       float64
+	Max       float64
+	Increment float64
+}
+
+// UnmarshalJSON : custom unmarshaller to handle legacy string granularites.
+func (pg *PriceGranularity) UnmarshalJSON(b []byte) error {
+	// We default to medium
+	if len(b) == 0 {
+		*pg = priceGranularityMed
+		return nil
+	}
+	// First check for legacy strings
+	var pgString string
+	err := json.Unmarshal(b, &pgString)
+	if err == nil {
+		*pg = PriceGranularityFromString(pgString)
+		if len(*pg) > 0 {
+			// Only exit if we matched something, else we try processing as custom granularity
+			// This way we error as expecting the new custom granularity standard.
+			return nil
+		}
+	}
+	// Not legacy, so we do a normal Unmarshal
+	gran := []GranularityRange{}
+	err = json.Unmarshal(b, &gran)
+	if err != nil {
+		return err
+	}
+	if len(gran) > 1 {
+		// We only need to loop over the ranges if we have more than one (inter-range checks)
+		precision := gran[0].Precision
+		var prevMax float64 = 0
+		for i, gr := range gran {
+			if gr.Precision != precision {
+				return errors.New("Price granularity error: precision not consistent across entires")
+			}
+			if gr.Max < prevMax {
+				return errors.New("Price granularity error: range list must be ordered with increasing \"max\"")
+			}
+			if gr.Min == 0.0 {
+				// Default min to be the previous max
+				// On the first entry, we will likely overwrite 0.0 with 0.0, which should be ok. Adding a conditional to skip likely won't save any processing time.
+				gr.Min = prevMax
+				gran[i].Min = prevMax
```

Comment:
Why are there two lines here? Both look to be doing the same thing.

Reply:
Yes, except `gr` is a copy I believe, so to have the change stick I need to change `gran[i]`. But the proper resolution is to just test `gran[i].Min` below rather than `gr.Min`. Fixing